### ISSUE TITLE
Tokens, where to start and stop

### ISSUE THREAD

**Issue by 1sra3l** [curlpipe/synoptic#3]
Thanks for the library, it saves so much time getting started!  I am using your library in a way you did not plan for, but allow.
I am using FLTK, and the backend is a buffer full of single characters.
```rs
fn example() {
    // test
}
```
Here is the FLTK version of the github highlighting above MOL
```
BBACCCCCCCAAAA
AAAAEEEEEEE
A
```
I currently use something like this:
```rs
let mut output:string = String::new();
//snip
for token in row {
    match token {
        Token::Start(kind) => {
            highlighting = true;
            match kind.as_str() {
                "keywords" => character = "B".to_string(),
                "comments" =>  character = "E".to_string(),
                //etc..
                _=> character = "A".to_string(),
            }
            output.push_str(character.as_str());// <-- here?
        }
        Token::Text(txt) => {
            let len = txt.chars().count();
            let ch = character.as_str();
            // `string.len()` == `text.len()`
            let string = ch.repeat(len);
            output.push_str(string.as_str());
        },
        Token::End(_) =>  {
            //output.push_str(character.as_str()); // <-- OR here?
            highlighting = false;
            character = "A".to_string();
        },
    }
}
```
Just not sure how to get it to be the correct length of the styling.
non-highlighted characters creep in, but  I also end up with an incorrect `len()` if I do not push a some character to the output string.
![image](https://user-images.githubusercontent.com/76438083/152026206-4a44e51e-9e9b-496b-9315-293baa834141.png)

That said, your library is quite intuitive and I really enjoy the thought you put into making it flexible!

**Comment by curlpipe** [comment]
Hi there! Thanks for using this library. 
Glad you find it useful. I felt that Rust's existing syntax highlighting libraries really lacked a low level and lightweight option, so I built this.

If I understand your problem correctly, I have a potential solution.

```rust
use synoptic::{Token, Highlighter};

fn main() {
    // Demonstration code
    let code = "fn example() {
    // test
}";
    // Create highlighter
    let mut hl = Highlighter::new();
    hl.add("fn", "keyword").unwrap();
    hl.add(r"([a-z][a-zA-Z0-9]*)\(", "function").unwrap();
    hl.add(r"//.*", "comment").unwrap();
    // Perform highlighting
    let highlighted = hl.run(code);
    // Form output
    let mut output = String::new();
    let mut character;
    for row in highlighted {
        // Reset highlighting on each row
        character = 'A';
        // Form this row
        for token in row {
            match token {
                Token::Start(kind) => {
                    character = match kind.as_str() {
                        "keyword" => 'B',
                        "function" => 'C',
                        "comment" => 'E',
                        _ => unreachable!(),
                    };
                }
                Token::Text(txt) => {
                    let len = txt.chars().count();
                    output.push_str(&character.to_string().repeat(len));
                }
                Token::End(_) => {
                    // Reset to default character 'A'
                    character = 'A';
                }
            }
        }
        // Add a newline to the end of the row
        output.push('\n');
    }
    // Print result and hope for the best ;)
    println!("{}", output);
}
```

Hopefully you can modify it to fit your needs.

Sounds like an interesting project. Let me know how it goes!

If you have any further questions, no matter how big or small, please do let me know. I'll be happy to answer.

**Comment by 1sra3l** [comment]
OH!  The `output.push('\n');` is what I missed!  :smile_cat:

Thanks so much!  I am converting old C++ programs I made using FLTK + FLUID, so I have the entire UI already.
I had a lexer I modified for that version, so having a library that already does what I want it to do is wonderful.  The Regex functionality makes it even easier!

It is part of a larger group of apps I am converting to use Rust, if you are interested I'll let you know when I get things working nicely by just posting the info on this closed bug.
